pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go

```diff
@@ -271,12 +293,61 @@ func (m *ManifestWorkController) applyOneManifest(
 	// TODO we should check the certain error.
 	// Use dynamic client when scheme cannot decode manifest or typed client cannot handle the object
 	if isDecodeError(result.Error) || isUnhandledError(result.Error) || isUnsupportedError(result.Error) {
-		result.Result, result.Changed, result.Error = m.applyUnstructured(ctx, unstructuredObj, gvr, recorder)
+		result.Result, result.Changed, result.Error = m.applyUnstructured(ctx, required, gvr, recorder)
 	}
 
 	return result
 }
 
+func (m *ManifestWorkController) serverSideApply(
+	ctx context.Context,
+	gvr schema.GroupVersionResource,
+	required *unstructured.Unstructured,
+	config *workapiv1.ServerSideApplyConfig,
+	recorder events.Recorder) (*unstructured.Unstructured, bool, error) {
+	force := false
+	fieldManager := "work-agent"
+
+	if config != nil {
+		force = config.Force
+		if len(config.FieldManager) > 0 {
+			fieldManager = config.FieldManager
+		}
+	}
+
+	existing, err := m.spokeDynamicClient.
+		Resource(gvr).
+		Namespace(required.GetNamespace()).
+		Get(ctx, required.GetName(), metav1.GetOptions{})
+
+	switch {
+	case err == nil:
+		// merge ownerref by existing. We need this to remove the ownerref of appliedManifestWork when user
+		// changes the deleteOption to Orphan.
+		modified := resourcemerge.BoolPtr(false)
+		existingOwners := existing.GetOwnerReferences()
+		resourcemerge.MergeOwnerRefs(modified, &existingOwners, required.GetOwnerReferences())
```

Comment:
combining client-side patch and SSA looks odd to me. also if we're merging this, we should also deal with the merging of corresponding entries of `.metadata.managedFields`

Reply:
In deleteOption feature, we need to ensure that the ownerrefs of appliedManifestwork should not exist. This is to deal with the case that user create a manifestwork and then update the manfiestwork's deleteOption to orphan. We need to get the existing owners on the resource, and ensure that owner of appliedManifestWork exists or not based on deleteOption. Since ownerrefs is a slice, I think we replace  the whole slice here.

Reply:
to be clarify, this code is not to generate a merge patch, but update the required object by changing the ownerrefs on it.

Reply:
I update this part. We should treat ownerrefs separately in server apply case, so it would not mess up with other field. Also add an integration test.